### ISSUE TITLE
Testing: Do(n't) test private functions

### ISSUE THREAD

**Issue by boennemann** [upfrontIO/guides#20]
Seen here: https://github.com/upfrontIO/livingdocs-framework/blob/feat-undo-redo-event/src/history/index.coffee#L71

@masone 

> In general: Don't test private functions. They should be implicitly tested by the public functions calling it. Could be a smell if you feel like you have to test them.
https://www.quora.com/Should-you-unit-test-private-methods-on-a-class

> In this case: When I understand the exports correctly, they're available on the module to the public. Either make them properly private or accept that they're public. There is no point in adding a comment stating they're private imho.

**Comment by boennemann** [comment]
The methods are exported on an internal module that itself exports a mixin-esque function.
Only the return value of the main export will ever end up in a public API. The private functions are indeed private, because it's not possible to call them directly when requiring `@livingdocs/framework`.

With that out of the way: I disagree completely with a rule that would disallow me to test private functions, in fact I would encourage everyone to make smaller logical bits of their code independent and to export them. That way it's possible to test every single code path in isolation with a profound unit test suite, without dependencies, and with tests that are running fast.

Of course it's necessary to test whether my public functions are coupled to them correctly, but I don't need that much of a coverage now. As soon as I know that they're coupled correctly, I'm trusting the lower level unit test suite to take care of all cases for me.

That's exactly the strategy I've applied to undo/redo. I've tested every single code path inside the history module itself, while inside the livingdoc I'm only testing if things are coupled correctly, and callable from the outside. Finally there is one "integration" run, where I'm quickly walking through the functionality, by calling only public user-facing apis, to test whether things are working on a very high level.

That way the more expensive, slower tests, with lots of dependencies and code running are reduced to a minimum, but they will fail as soon as the coupling to the rest of the code fails. My unit-tests can then take care of every single bit of functionality in detail, without taking any resources to run.
